library/Booster/Syntax/Json/Internalise.hs

```diff
@@ -269,41 +311,111 @@ internaliseTerm ::
+  ensure nothing is circular or ambiguous.
+
+  Throws a PatternError when any of the top-level patterns is a term
+  or an And with a term (must be analysed beforehand).
+-}
+internalisePredicates ::
+    Flag "alias" ->
+    Flag "subsorts" ->
+    Maybe [Syntax.Id] ->
+    KoreDefinition ->
+    [Syntax.KorePattern] ->
+    Except PatternError InternalisedPredicates
+internalisePredicates allowAlias checkSubsorts sortVars definition ps = do
+    internalised <-
+        concatMapM (internalisePred allowAlias checkSubsorts sortVars definition) $
+            concatMap explodeAnd ps
+
+    let (substitution, moreEquations) =
+            mkSubstitution [s | s@SubstitutionPred{} <- internalised]
+
+    pure
+        InternalisedPredicates
+            { boolPredicates = Set.fromList $ [p | BoolPred p <- internalised] <> moreEquations
+            , ceilPredicates = Set.fromList $ [p | CeilPred p <- internalised]
+            , substitution
+            , unsupported = [p | UnsupportedPred p <- internalised]
+            }
+  where
+    mkSubstitution ::
+        [InternalisedPredicate] -> (Map Internal.Variable Internal.Term, [Internal.Predicate])
+    mkSubstitution initialSubst =
+        let substMap, duplicates :: Map Internal.Variable [Internal.Term]
+            (substMap, duplicates) =
+                Map.partition ((== 1) . length) $
+                    Map.fromListWith (<>) [(v, [t]) | SubstitutionPred v t <- initialSubst]
+            equations =
+                [mkEq v t | (v, ts) <- Map.assocs duplicates, t <- ts]
+         in execState breakCycles (Map.map head substMap, equations)
+
+    breakCycles :: State (Map Internal.Variable Internal.Term, [Internal.Predicate]) ()
+    breakCycles = do
+        assocs <- gets (Map.assocs . fst)
+        let sccs =
+                stronglyConnComp [(v, v, Set.toList (freeVariables t)) | (v, t) <- assocs]
+            cycleNodes = Set.fromList [v | CyclicSCC (v : _vs) <- sccs]
+        if null cycleNodes
+            then pure ()
+            else do
+                modify $ \(m, eqs) ->
+                    ( m `Map.withoutKeys` cycleNodes
+                    , eqs <> (map (uncurry mkEq) $ Map.assocs $ m `Map.restrictKeys` cycleNodes)
+                    )
+                breakCycles
+
+mkEq :: Internal.Variable -> Internal.Term -> Internal.Predicate
+mkEq x t = Internal.Predicate $ case sortOfTerm t of
+    Internal.SortInt -> Internal.EqualsInt (Internal.Var x) t
+    Internal.SortBool -> Internal.EqualsBool (Internal.Var x) t
+    otherSort -> Internal.EqualsK (Internal.KSeq otherSort (Internal.Var x)) (Internal.KSeq otherSort t)
+
+internalisePred ::
     Flag "alias" ->
     Flag "subsorts" ->
     Maybe [Syntax.Id] ->
     KoreDefinition ->
     Syntax.KorePattern ->
-    Except PatternError InternalisedPredicate
-internalisePredicate allowAlias checkSubsorts sortVars definition@KoreDefinition{sorts} pat = case pat of
+    Except PatternError [InternalisedPredicate]
```

Comment:
not sure if we need to return a list here. since we don't handle #and under a #not and we always explode ands before calling this function (afaik), this always seems to return a singleton

Reply:
The case of `#And` requires the list, and we need it in the recursion under `#Not`.
Of course we could try to call `explodeAnd` before the recursion and re-assemble the (singleton) results afterwards in the right way (then assuming that we never find `#And` at the top in `internalisePred`).
I can maybe make a follow-up PR to change this (and handle  some more cases inside `#Not` as commented below).